proxy_agent/src/key_keeper.rs

```diff
@@ -147,72 +138,69 @@ fn poll_secure_channel_status(
                         None => err_string,
                     }
                 );
-                unsafe {
-                    *STATUS_MESSAGE = message.to_string();
-                }
+                key_keeper_wrapper::set_status_message(shared_state.clone(), message.to_string());
                 logger::write_warning(message);
                 continue;
             }
         };
-        let mut guid;
-        match &status.keyGuid {
-            Some(id) => guid = id.to_string(),
-            None => guid = String::new(),
-        }
-
         logger::write_information(format!("Got key status successfully: {}.", status));
 
         let wireserver_rule_id = status.get_wireserver_rule_id();
-        let imds_rule_id = status.get_imds_rule_id();
-        unsafe {
-            if wireserver_rule_id != *WIRESERVER_RULE_ID {
-                logger::write_warning(format!(
-                    "Wireserver rule id changed from {} to {}.",
-                    *WIRESERVER_RULE_ID, wireserver_rule_id
-                ));
-                *WIRESERVER_RULE_ID = wireserver_rule_id.to_string();
-                proxy_authentication::set_wireserver_rules(status.get_wireserver_rules());
-            }
+        let imds_rule_id: String = status.get_imds_rule_id();
+        let current_wire_server_rule_id =
+            key_keeper_wrapper::get_wireserver_rule_id(shared_state.clone());
+        if wireserver_rule_id != current_wire_server_rule_id {
```

Comment:
There's a race condition here where you're reading the current rule, releasing the lock, comparing it, and then re-acquiring the lock and updating it based on the comparison. Another thread could potentially set it such that `wireserver_rule_id != current_wire_server_rule_id` after you read `wireserver_rule_id == current_wire_server_rule_id` and you won't update the rule like you expect.

You can either unconditionally set the value, or (if you want to stick with your wrapper approach) add a wrapper that holds the lock across the read and conditional write.